File: leetcode/problems/110.balanced-binary-tree.java
/*
 * @lc app=leetcode id=110 lang=java
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start

//  Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    boolean result = true;

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }

        dfs(root, 0);

        return this.result;
    }

    private int dfs(TreeNode node, int depth) {
        if (node == null) {
            return depth;
        }

        int left = dfs(node.left, depth + 1);
        int right = dfs(node.right, depth + 1);

        if (Math.abs(left - right) > 1) {
            this.result = false;
        }

        return Math.max(left, right);
    }
}
// @lc code=end

File: leetcode/problems/110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#


# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    result = True

    def isBalanced(self, root: TreeNode | None) -> bool:
        if not root:
            return True

        def dfs(node, depth):
            if not node:
                return depth

            left = dfs(node.left, depth + 1)
            right = dfs(node.right, depth + 1)

            if abs(left - right) > 1:
                self.result = False

            return max(left, right)

        dfs(root, 0)

        return self.result


# @lc code=end
